### PR TITLE
Make NamingPreferences message-pack serializable

### PR DESCRIPTION
**Pull request by tmat**
Implements https://github.com/dotnet/roslyn/issues/60849

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs [hunk 1]
```diff
@@ -334,44 +346,72 @@ private interface ISymbolMatcher
...
+            private const int KindBits = 6;
+
+            private enum SymbolCategory
+            {
+                Other = 1,
+                Type = 2,
+                Method = 3,
+            }
+
+            // highest 2 bits for category, 6 bits for kind
```
Comment (dibarbet) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs:361]: not super familiar with message pack serialization - is it not possible to just have the type and kind as properties?  Or was there a perf issue / message size problem if we don't pack this into a byte?

**Review by dibarbet — APPROVED** [review]

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/NamingStyle.cs [hunk 2]
```diff
@@ -19,18 +20,34 @@
...
+    [DataContract]
+    internal readonly partial record struct NamingStyle : IObjectWritable
     {
-        public Guid ID { get; }
-        public string Name { get; }
-        public string Prefix { get; }
-        public string Suffix { get; }
-        public string WordSeparator { get; }
-        public Capitalization CapitalizationScheme { get; }
+        [DataMember(Order = 0)]
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/NamingStyle.cs:26]: ❔ Do we not have anything equivalent to `StructLayout.Sequential`? These attributes seem like unnecessary noise.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SerializableNamingRule.cs [hunk 3]
```diff
@@ -5,16 +5,23 @@
...
-        public Guid NamingStyleID;
-        public ReportDiagnostic EnforcementLevel;
+        [DataMember(Order = 0)]
+        public Guid SymbolSpecificationID { get; init; }
+
+        [DataMember(Order = 1)]
+        public Guid NamingStyleID { get; init; }
+
+        [DataMember(Order = 2)]
+        public ReportDiagnostic EnforcementLevel { get; init; }
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SerializableNamingRule.cs:24]: ❔ Why not `{ get; }`?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs (hunk 1 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SymbolSpecification.cs:361]: It seems like unnecessary complexity, with a slight risk of regression if a future maintainer introduces an error in the shifting/masking.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SerializableNamingRule.cs (hunk 3 above)
Comment (tmat) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/NamingStyles/Serialization/SerializableNamingRule.cs:24]: `SerializableNamingRule` does not have an explicit constructor. `init;` is to make this object immutable, is not related to message-pack. An alternative is to write an explicit constructor.

**Pull request merged**